reference-implementation/to-upstream-wpts/writable-streams/close.js

```diff
@@ -61,108 +62,72 @@ promise_test(t => {
+    writer.close();
+    assert_equals(writer.desiredSize, 1, 'desiredSize should be still 1');
 
-      writer.ready.then(t.step_func(v => {
-        assert_equals(v, undefined, 'ready promise was fulfilled with undefined');
-        t.done();
-      }));
-    }));
-  }), 0);
-}, 'If close is called on a WritableStream in writable state, ready will return a fulfilled promise');
+    return writer.ready.then(v => {
+      assert_equals(v, undefined, 'ready promise should be fulfilled with undefined');
+      assert_array_equals(ws.events, ['close'], 'write and abort should not be called');
+    });
+  });
+}, 'when close is called on a WritableStream in writable state, ready should return a fulfilled promise');
 
-async_test(t => {
-  const ws = new WritableStream({
+promise_test(t => {
+  const ws = recordingWritableStream({
     write() {
       return new Promise(() => {});
-    },
-    abort() {
-      t.step(() => {
-        assert_unreached('Unexpected abort call');
-      });
     }
   });
 
   const writer = ws.getWriter();
 
-  // Wait for ws to start.
-  setTimeout(t.step_func(() => {
+  return writer.ready.then(() => {
     writer.write('a');
 
     assert_equals(writer.desiredSize, 0, 'desiredSize should be 0');
 
-    let closeCalled = false;
-
-    writer.ready.then(t.step_func(v => {
-      if (closeCalled === false) {
-        assert_unreached('ready fulfilled before writer.close()');
-        return;
-      }
-
-      assert_equals(v, undefined, 'ready promise was fulfilled with undefined');
-      t.done();
-    }));
-
-    setTimeout(t.step_func(() => {
-      writer.close();
-      closeCalled = true;
-    }), 100);
-  }), 0);
-}, 'If close is called on a WritableStream in waiting state, ready promise will fulfill');
+    let calledClose = false;
+    return Promise.all([
+      writer.ready.then(v => {
+        assert_equals(v, undefined, 'ready promise should be fulfilled with undefined');
+        assert_true(calledClose, 'ready should not be fulfilled before writer.close() is called');
+        assert_array_equals(ws.events, ['write', 'a'], 'sink abort() should not be called');
+      }),
+      delay(100).then(() => {
+        writer.close();
+        calledClose = true;
+      })
+    ]);
+  });
+}, 'when close is called on a WritableStream in waiting state, ready promise should be fulfilled');
 
-async_test(t => {
-  let readyFulfilledAlready = false;
-  const ws = new WritableStream({
-    abort() {
-      t.step(() => {
-        assert_unreached('Unexpected abort call');
-      });
-    },
+promise_test(t => {
+  let asyncCloseFinished = false;
+  const ws = recordingWritableStream({
     close() {
-      return new Promise(resolve => {
-        setTimeout(t.step_func(() => {
-          t.ok(readyFulfilledAlready, 'ready should have fulfilled already');
-          resolve();
-        }), 50);
-      });
+      return delay(50).then(() => asyncCloseFinished = true);
```

Comment:
I reversed the logic so that I wouldn't need an assert inside the sink method. Simplifying the test was just a happy side-effect.